docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+- *Guaranteed* pods are pods where every container, for both memory and CPU, has the same request and limit, and request is non-zero for every container. These pods ask for a well-defined amount of resources and are therefore considered top-priority.
+- All other pods are *Burstable* - middle priority pods that have some form of minimal resource guarantee, but can use more resources when available.
+- A container than specifies a limit of 0 should be considered invalid.
+
+Note that the policy of classifying pods is for the most part orthogonal to the admission policy (described next), and can be easily tweaked.
+
+### Admission Policy
+- Pods will be admitted by Kubelet based on the sum of requests of its containers. The Kubelet will ensure that sum of requests of all containers (over all pods) is within the system’s resource (for both memory and CPU).
+- One big guiding principle is that Best-Effort pods should be “invisible” to all other pods. The main purpose of Best-Effort pods is to utilize resources not used by higher-tier pods. Consequently, Guaranteed and Burstable pods will pre-empt Best-Effort pods if necessary.
+  - Guaranteed pods have a strong guarantee that they will not be pre-empted by any other pods.
+  - Because of implementation challenges, Burstable pods might be pre-empted by Best-Effort pods (but the Kubelet will try to prevent this).
+  - One important instance pre-emption takes into effect is when a node is running the maximum number of pods (currently 40). Suppose that 40 best-effort pods are running on a node, and a guaranteed pod wants to run on the node. The oldest running best-effort pod will be removed and the guaranteed pod will be admitted (if the guaranteed pod’s resources fit in the machine’s capacity).
+
+## Alternatives Considered
+
+Designing a simple but flexible quality of service framework is very challenging because there are many possible applications that can use tiered pods in different ways. One alternative is to have user-specified numerical priorities that guide Kubelet on which tasks to kill (if the node runs out of memory, lower priority tasks will be killed). A strict hierarchy of user-specified numerical priorities is not desirable because: a) Achieved behavior would be emergent based on how users assigned priorities to their pods. No particular SLO could be delivered by the system, and usage would be subject to gaming if not restricted administratively, b) Changes to desired priority bands would require changes to all user pod configurations.
+
+## Implementation
+
+Add kubelet/qos/policy.go
+- Decides which tier a pod is in (based on the tier policy above)
+- Decides what OOM score all processes in a pod should get
+- Decides how many CPU shares a container should get
+
+Scheduler Changes (scheduler/algorithm/predicates.go)
+- Use requests instead of limits in CheckPodsExceedingCapacity and PodFitsResources
+- Modify pod admission so that best-effort pods are pre-empted when necessary

```

Comment:
Removing preemption!

In the (outdated) proposal, best effort pods were pods where all the requests were 0 (for both CPU and memory). My thinking was that best-effort pods should be "invisible" to other pods. So if there are MaxPods best-effort pods running on a node, and a guaranteed pod wants to run on the pod, Kubelet should evict the best-effort pod and admit the guaranteed pod.

I think we will need to look a bit more into how we classify pods as best-effort. I'm planning to post my thoughts on this in another PR :)